src/rnnlm/rnnlm-utils.cc

```diff
@@ -0,0 +1,145 @@
+#include "rnnlm/rnnlm-utils.h"
+
+namespace kaldi {
+namespace rnnlm {
+
+vector<string> SplitByWhiteSpace(const string &line) {
+  std::stringstream ss(line);
+  vector<string> ans;
+  string word;
+  while (ss >> word) {
+    ans.push_back(word);
+  }
+  return ans;
+}
+
+unordered_map<string, int> ReadWordlist(string filename) {
+  unordered_map<string, int> ans;
+  ifstream ifile(filename.c_str());
+  string word;
+  int id;
+
+  while (ifile >> word >> id) {
+    ans[word] = id;
+  }
+  return ans;
+}
+
+NnetExample GetEgsFromSent(const vector<int>& word_ids_in, int input_dim,
+                           const vector<int>& word_ids_out, int output_dim) {
+  SparseMatrix<BaseFloat> input_frames(word_ids_in.size(), input_dim);
+
+  for (int j = 0; j < word_ids_in.size(); j++) {
+    vector<std::pair<MatrixIndexT, BaseFloat> > pairs;
+    pairs.push_back(std::make_pair(word_ids_in[j], 1.0));
+    SparseVector<BaseFloat> v(input_dim, pairs);
+    input_frames.SetRow(j, v);
+  }
+
+  NnetExample eg;
+  eg.io.push_back(nnet3::NnetIo("input", 0, input_frames));
```

Comment:
[Important]
I notice that the egs are of variable length, equal to the sentence length.
This is going to be very inefficient when compiling computations, if the sentences are of variable length.
However, I am in the process of making some changes to the egs-merging that should make this much less of a problem-- it will enable us to batch up sentences that are the same length into minibatches, which will mean that the egs have a more regular structure and the structure is more frequently repeated.  I am also making the nnet3 compilation process much more efficient (on egs with regular structure).
This should be checked in in 2 or 3 weeks if all goes to plan.
In your logs keep track of the "total GPU time" reported, versus the total time taken you'll see in the Accounting:" line.  If the total GPU time is much less than the total, then you are taking time not on GPU (e.g. for compilation, or something like that).  Point me to one of your logs.